src/math/window.c

```diff
@@ -0,0 +1,141 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>
+
+#include <sof/audio/format.h>
+#include <sof/math/decibels.h>
+#include <sof/math/log.h>
+#include <sof/math/trig.h>
+#include <sof/math/window.h>
+#include <errno.h>
+#include <stdint.h>
+
+#define WIN_ONE_Q15 INT16_MAX
+#define WIN_ONE_Q31 INT32_MAX
+#define WIN_05_Q31 Q_CONVERT_FLOAT(0.5, 31)
+
+#define WIN_TWO_PI_Q28 Q_CONVERT_FLOAT(6.2831853072, 28)
+
+#define WIN_085_Q31 Q_CONVERT_FLOAT(0.85, 31)
+#define WIN_LOG_2POW31_Q26 Q_CONVERT_FLOAT(21.4875625974, 26)
+
+/* Exact
+ * #define WIN_HAMMING_A0_Q30 Q_CONVERT_FLOAT(25.0 / 46.0, 30)
+ * #define WIN_HAMMING_A1_Q30 Q_CONVERT_FLOAT(1 - 25.0 / 46.0, 30)
+ */
+
+/* Common approximations to match e.g. Octave */
+#define WIN_HAMMING_A0_Q30 Q_CONVERT_FLOAT(0.54, 30)
+#define WIN_HAMMING_A1_Q30 Q_CONVERT_FLOAT(0.46, 30)
+
+/**
+ * \brief Return matrix of Q1.15 ones
+ * \param[in,out] x[][] Output two dimensional array
+ * \param[in] rows Number of rows
+ * \param[in] columns Number of columns
+ *
+ */
+
+/* TODO: Decide how to present one and two dimensional arrays, for matrix
+ * operations library. pointer arithmetic like below vs. x[][]
+ */
+static void mat_ones_16b(int16_t *x, int rows, int columns)
+{
+	int i;
+	int j;
+
+	for (i = 0; i < rows; i++) {
+		for (j = 0; j < columns; j++)
+			*(x + columns * i + j) = WIN_ONE_Q15;
+	}
+}
+
+/**
+ * \brief Return rectangular window, simply values of one
+ * \param[in,out]  win  Output vector with coefficients
+ * \param[in]  length  Length of coefficients vector
+ */
+void win_rectangular_16b(int16_t win[], int length)
+{
+	mat_ones_16b(win, length, 1);
+}
+
+/**
+ * \brief Calculate Blackman window function, reference
+ * https://en.wikipedia.org/wiki/Window_function#Blackman_window
+
+ * \param[in,out]  win  Output vector with coefficients
+ * \param[in]  length  Length of coefficients vector
+ * \param[in]  a0      Parameter for window shape, use e.g. 0.42 as Q1.15
+ */
+void win_blackman_16b(int16_t win[], int length, int16_t a0)
+{
+	const int16_t a1 = Q_CONVERT_FLOAT(0.5, 15);
+	int32_t inv_length;
+	int32_t val;
+	int32_t a;
+	int16_t alpha;
+	int16_t a2;
+	int16_t c1;
+	int16_t c2;
+	int n;
+
+	alpha = WIN_ONE_Q15 - 2 * a0; /* Q1.15 */
+	a2 = alpha / 2;
+	a = WIN_TWO_PI_Q28 / (length - 1); /* Q4.28 */
+	inv_length = WIN_ONE_Q31 / length;
+
+	for (n = 0; n < length; n++) {
+		c1 = cos_fixed_16b(a * n);
+		c2 = cos_fixed_16b(2 * Q_MULTSR_32X32((int64_t)a, inv_length, 28, 31, 28));
```

Comment:
a *n, n is missing from the calculation?

Reply:
Yes, good catch!

Reply:
Done, the RMS error decreased from 2.9 LSB to 0.67 LSB.